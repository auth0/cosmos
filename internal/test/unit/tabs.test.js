import React from 'react'
import { shallow } from 'enzyme'

import Tabs from '@auth0/cosmos/molecules/tabs/tabs'

function tabsFactory() {
  const content = {
    first: <div className="content-1" />,
    second: <div className="content-2" />,
    third: <div className="content-3" />
  }

  const generator = (index = 0, onSelect = () => {}) =>
    shallow(
      <Tabs selected={index} onSelect={onSelect}>
        <Tabs.Tab title="Title 1">{content.first}</Tabs.Tab>
        <Tabs.Tab title="Title 2">{content.second}</Tabs.Tab>
        <Tabs.Tab title="Title 3">{content.third}</Tabs.Tab>
      </Tabs>
    )

  return { generator, content }
}

describe('Tabs', () => {
  it('renders only selected tab content', () => {
    jest.mock('../../../core/components/_helpers/uniqueId', () => () => 'abcdef1234')
    const { generator: wrapper } = tabsFactory()

    expect(wrapper).toMatchSnapshot()
  })

  it('only render one tab title as selected', () => {
    const { generator } = tabsFactory()

    // Create three <Tabs> instances with different tabs selected
    const [first, second, third] = [generator(0), generator(1), generator(2)]

    expect(first).toMatchSnapshot()
    expect(second).toMatchSnapshot()
    expect(third).toMatchSnapshot()
  })

  it('onSelect is called on tab title click', () => {
    const { generator } = tabsFactory()
    const selectedHandler = jest.fn()

    const wrapper = generator(0, selectedHandler)
    const unSelectedTab = wrapper
      .find(Tabs.TabLink)
      .filter({ 'aria-selected': false })
      .first()

    unSelectedTab.simulate('click')

    expect(selectedHandler).toHaveBeenCalled()
  })

  it('renders linked accessibility identifiers', () => {
    const { generator } = tabsFactory()
    const testableIndexes = [0, 1, 2]

    testableIndexes.forEach(index => {
      const tabs = generator(index)

      const [tabList, activeTabPanel] = tabs.children()

      const getPropFromTabLink = propName =>
        tabList.props.children.map(link => link.props.children.props[propName])

      const getPropFromTabPane = (propName, removeValue) => {
        const value = activeTabPanel.props[propName]
        return removeValue ? value.replace(removeValue, '') : value
      }

      const tabLinkIds = getPropFromTabLink('id')
      const tabLinkAriaControls = getPropFromTabLink('aria-controls')

      expect(tabLinkIds[index]).toEqual(getPropFromTabPane('id', '-tab'))
      expect(tabLinkIds[index]).toEqual(getPropFromTabPane('aria-labelledby', '-tab'))
      expect(tabLinkAriaControls[index]).toEqual(getPropFromTabPane('id'))
    })
  })
})
