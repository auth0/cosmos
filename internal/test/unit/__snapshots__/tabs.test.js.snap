// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs only render one tab title as selected 1`] = `
<styled.div
  data-cosmos-key="tabs"
  id="-m0ck3d"
>
  <styled.ul
    data-cosmos-key="tabs.list"
    role="tablist"
  >
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-0"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-0-tab"
        aria-selected={true}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="0"
        type="button"
      />
    </styled.li>
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-1"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-1-tab"
        aria-selected={false}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="-1"
        type="button"
      />
    </styled.li>
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-2"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-2-tab"
        aria-selected={false}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="-1"
        type="button"
      />
    </styled.li>
  </styled.ul>
  <styled.div
    aria-labelledby="tabs-m0ck3d-0-tab"
    data-cosmos-key="tabs.item"
    id="tabs-m0ck3d-0-tab"
    key=".0"
    role="tabpanel"
    title="Title 1"
  >
    <div
      className="content-1"
    />
  </styled.div>
</styled.div>
`;

exports[`Tabs only render one tab title as selected 2`] = `
<styled.div
  data-cosmos-key="tabs"
  id="-m0ck3d"
>
  <styled.ul
    data-cosmos-key="tabs.list"
    role="tablist"
  >
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-0"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-0-tab"
        aria-selected={false}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="-1"
        type="button"
      />
    </styled.li>
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-1"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-1-tab"
        aria-selected={true}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="0"
        type="button"
      />
    </styled.li>
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-2"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-2-tab"
        aria-selected={false}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="-1"
        type="button"
      />
    </styled.li>
  </styled.ul>
  <styled.div
    aria-labelledby="tabs-m0ck3d-1-tab"
    data-cosmos-key="tabs.item"
    id="tabs-m0ck3d-1-tab"
    key=".1"
    role="tabpanel"
    title="Title 2"
  >
    <div
      className="content-2"
    />
  </styled.div>
</styled.div>
`;

exports[`Tabs only render one tab title as selected 3`] = `
<styled.div
  data-cosmos-key="tabs"
  id="-m0ck3d"
>
  <styled.ul
    data-cosmos-key="tabs.list"
    role="tablist"
  >
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-0"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-0-tab"
        aria-selected={false}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="-1"
        type="button"
      />
    </styled.li>
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-1"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-1-tab"
        aria-selected={false}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="-1"
        type="button"
      />
    </styled.li>
    <styled.li
      data-cosmos-key="tabs.list-item"
      key="tabs-m0ck3d-2"
      role="presentation"
    >
      <styled.button
        aria-controls="tabs-m0ck3d-2-tab"
        aria-selected={true}
        data-cosmos-key="tabs.link"
        id="tabs-m0ck3d-2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex="0"
        type="button"
      />
    </styled.li>
  </styled.ul>
  <styled.div
    aria-labelledby="tabs-m0ck3d-2-tab"
    data-cosmos-key="tabs.item"
    id="tabs-m0ck3d-2-tab"
    key=".2"
    role="tabpanel"
    title="Title 3"
  >
    <div
      className="content-3"
    />
  </styled.div>
</styled.div>
`;

exports[`Tabs renders only selected tab content 1`] = `[Function]`;
